### PR TITLE
chore: migrate media-library components to Typescript

### PR DESCRIPTION
**Pull request by simotae14**
### What does it do?

- Migrate media-library plugin admin to Typescript
  - migrate all the components to Typescript

based on the work made in the draft PR https://github.com/strapi/strapi/pull/20867

### Files to migrate?

- [ ] admin
  - [ ] src
    - [x] components
      - [x] AssetCard
        - [x] tests
          - [x] AssetCardBase.test.jsx
          - [x] DocAssetCard.test.jsx
          - [x] ImageAssetCard.test.jsx 
        - [x] AssetCard.jsx
        - [x] AssetCardBase.jsx
        - [x] AudioAssetCard.jsx
        - [x] AudioPreview.jsx
        - [x] DocAssetCard.jsx
        - [x] ImageAssetCard.jsx
        - [x] UploadingAssetCard.jsx
        - [x] VideoAssetCard.jsx
        - [x] VideoPreview.jsx 
      - [x] AssetDialog
        - [x] BrowseStep
          - [x] PaginationFooter
            - [x] components.jsx
            - [x] index.jsx
            - [x] Pagination.jsx
            - [x] PaginationContext.jsx  
        - [x] SearchAsset
          - [x] tests
            - [x] index.test.jsx
          - [x] index.jsx
        - [x] tests
          - [x] __snapshots__
          - [x] index.test.jsx
        - [x] utils
          - [x] tests
            - [x] isSelectable.test.js
          - [x] isSelectable.js     
        - [x] Filters.jsx
        - [x] index.jsx
        - [x] PageSize.jsx 
        - [x] SelectedStep
          - [x] index.jsx
        - [x] tests
          - [x] AssetDialog.test.jsx
        - [x] DialogFooter.jsx
        - [x] index.jsx  
      - [x] AssetGridList
        - [x] tests
          - [x] __snapshots__
          - [x] AssetGridList.test.jsx 
        - [x] Draggable.jsx
        - [x] index.jsx 
      - [x] Breadcrumbs
        - [x] tests
          - [x] __snapshots__
          - [x] index.test.jsx 
        - [x] Breadcrumbs.jsx
        - [x] CrumbSimpleMenuAsync.jsx
        - [x] index.jsx
      - [x] BulkMoveDialog
        - [x] tests
          - [x] __snapshots__
          - [x] BulkMoveDialog.test.jsx 
        - [x] BulkMoveDialog.jsx
        - [x] index.jsx 
      - [x] ContextInfo
        - [x] tests
          - [x] __snapshots__
          - [x] ContextInfo.test.jsx 
        - [x] ContextInfo.jsx
        - [x] index.jsx 
      - [x] CopyLinkButton
        - [x] index.jsx 
      - [x] EditAssetDialog
        - [x] PreviewBox
          - [x] AssetPreview.jsx
          - [x] components.jsx
          - [x] CroppingActions.jsx
          - [x] index.jsx
        - [x] tests
          - [x] __snapshots__
          - [x] EditAssetDialog.test.jsx
          - [x] index.test.jsx
          - [x] RemoveAssetDialog.test.jsx
        - [x] DialogHeader.jsx
        - [x] index.jsx
        - [x] RemoveAssetDialog.jsx
        - [x] ReplaceMediaButton.jsx     
      - [x] EditFolderDialog
        - [x] ModalHeader
          - [x] index.jsx
          - [x] ModalHeader.jsx
        - [x] tests
          - [x]  __snapshots__
          - [x] EditFolderDialog.test.jsx
        - [x] EditFolderDialog.jsx
        - [x] index.jsx
        - [x] RemoveFolderDialog.jsx
      - [x] EmptyAssets
        - [x] EmptyAssetGrid.jsx
        - [x] index.jsx
      - [x] FilterList
        - [x] tests
          - [x]  __snapshots__
          - [x] index.test.js
        - [x] FilterTag.jsx
        - [x] index.jsx  
      - [x] FilterPopover
        - [x] utils
          - [x] getFilterList.js
        - [x] FilterValueInput.jsx
        - [x] index.jsx 
      - [x] FolderCard
        - [x] contexts
        - [x] FolderCard
          - [x] FolderCard.jsx
          - [x] index.jsx 
        - [x] FolderCardBody
          - [x] FolderCardBody.jsx
          - [x] index.jsx
        - [x] FolderCardBodyAction
          - [x] index.js 
        - [x] FolderCardCheckbox
          - [x] FolderCardCheckbox.jsx
          - [x] index.jsx
        - [x] hooks
          - [x] tests
            - [x] useId.test.js
          - [x] useId.js
        - [x] tests
          - [x] __snapshots__
          - [x] FolderCard.test.jsx 
        - [x] index.jsx
      - [x] FolderGridList
        - [x] tests
          - [x] __snapshots__
          - [x] FolderGridList.test.jsx
        - [x] FolderGridList.jsx
        - [x] index.jsx 
      - [x] MediaLibraryDialog
        - [x] index.jsx 
      - [x] MediaLibraryInput
        - [x] Carousel
          - [x] tests
            - [x] CarouselAssets.test.jsx
          - [x] CarouselAsset.jsx
          - [x] CarouselAssetActions.jsx
          - [x] CarouselAssets.jsx
          - [x] EmptyStateAsset.jsx 
        - [x] tests
          - [x] MediaLibraryInput.test.jsx 
        - [x] index.jsx
      - [x] SelectTree
        - [x] tests
          - [x] __snapshots__
          - [x] SelectTree.test.jsx
        - [x] utils
          - [x] tests
            - [x] __snapshots__
            - [x] flattenTree.test.js
            - [x] getOpenValues.test.js
            - [x] getValuesToClose.test.js
          - [x] flattenTree.js
          - [x] getOpenValues.js
          - [x] getValuesToClose.js
        - [x] index.jsx
        - [x] Option.jsx
        - [x] SelectTree.jsx
        - [x] SelectTree.stories.mdx 
      - [x] SortPicker
        - [x] index.jsx
      - [x] TableList
        - [x] tests
          - [x] __snapshots__
          - [x] CellContent.test.jsx
          - [x] TableList.test.jsx
          - [x] TableRows.test.jsx
        - [x] CellContent.jsx
        - [x] index.jsx
        - [x] PreviewCell.jsx
        - [x] TableRows.jsx
      - [x] UploadAssetDialog
        - [x] AddAssetStep
          - [x] tests
            - [x] __snapshots__
            - [x] FromComputerForm.test.jsx
          - [x] AddAssetStep.jsx
          - [x] FromComputerForm.jsx
          - [x] FromUrlForm.jsx 
        - [x] PendingAssetStep
          - [x] tests
            - [x] __snapshots__
            - [x] PendingAssetStep.test.jsx
          - [x] PendingAssetStep.jsx 
        - [x] tests
          - [x] UploadAssetDialog.test.jsx
        - [x] UploadAssetDialog.jsx 
      - [x] UploadProgress
        - [x] tests
          - [x] index.test.jsx 
        - [x] index.jsx
    - [x] hooks
      - [x] __mocks
        - [x] useAssets.js
        - [x] useBulkMove.js
        - [x] useBulkRemove.js
        - [x] useConfig.js
        - [x] useEditFolder.js
        - [x] useFolder.js
        - [x] useFolders.js
        - [x] useFolderStructure.js
        - [x] useMediaLibraryPermissions.js
        - [x] useModalQueryParams.js 
      - [x] tests
        - [x] useAssets.test.js
        - [x] useBulkMove.test.js
        - [x] useBulkRemove.test.js
        - [x] useClipboard.test.ts
        - [x] useConfig.test.js
        - [x] useEditFolder.test.js
        - [x] useFolder.test.js
        - [x] useFolders.test.js
        - [x] useFolderStructure.test.js
        - [x] useModalQueryParams.test.js
        - [x] usePersistentState.test.ts
        - [x] useRemoveAsset.test.js
        - [x] useSelectionState.test.ts 
      - [x] utils
        - [x] tests
          - [x] rename-keys.test.js 
        - [x] rename-keys.js
      - [x] useAssets.js
      - [x] useBulkMove.js
      - [x] useBulkRemove.js
      - [x] useClipboard.ts
      - [x] useConfig.js
      - [x] useCropImg.js
      - [x] useEditAsset.js
      - [x] useEditFolder.js
      - [x] useFolder.js
      - [x] useFolders.js
      - [x] useFolderStructure.js
      - [x] useMediaLibraryPermissions.js
      - [x] useModalQueryParams.js
      - [x] usePersistentState.ts
      - [x] useRemoveAsset.js
      - [x] useSelectionState.ts
      - [x] useUpload.js 
    - [ ] pages
      - [ ] App
        - [ ] components
          - [ ] tests
            - [ ] EmptyOrNoPermissions.test.jsx
          - [ ] BulkActions.jsx
          - [ ] BulkDelete.jsx
          - [ ] BulkMoveButton.jsx
          - [ ] EmptyOrNoPermissions.jsx
          - [ ] Filters.jsx
          - [ ] Header.jsx
        - [ ] ConfigureTheView
          - [ ] components
            - [ ] tests
              - [ ] __snapshots 
            - [ ] Settings.jsx
          - [ ] state
            - [ ] tests
              - [ ] init.test.js
              - [ ] reducer.test.js
            - [ ] actions.js
            - [ ] actionTypes.js
            - [ ] init.js
            - [ ] reducer.js 
          - [ ] tests
            - [ ] __snapshots
            - [ ] ConfigureTheView.test.jsx
          - [ ] index.jsx
        - [ ] MediaLibrary
          - [ ] components
            - [ ] tests
              - [ ] __snapshots__
              - [ ] BulkActions.test.jsx
              - [ ] BulkDeleteButton.test.jsx
              - [ ] BulkMoveButton.test.jsx
              - [ ] EmptyOrNoPermissions.test.jsx
              - [ ] Header.test.jsx
            - [ ] BulkActions.jsx
            - [ ] BulkDeleteButton.jsx
            - [ ] BulkMoveButton.jsx
            - [ ] EmptyOrNoPermissions.jsx
            - [ ] Filters.jsx
            - [ ] Header.jsx
          - [ ] tests
            - [ ] MediaLibrary.test.jsx
          - [ ] index.jsx
        - [ ] index.jsx 
      - [ ] SettingsPage
        - [ ] tests
          - [ ] index.test.jsx
          - [ ] reducer.test.js 
        - [ ] index.jsx
        - [ ] init.js
        - [ ] reducer.js
    - [x] translations
      - [x] tests
        - [x] plural.test.js
    - [x] utils
      - [x] tests
        - [x] appendSearchParamsToUrl.test.js
        - [x] containsAssetFilter.test.js
        - [x] downloadFile.test.js
        - [x] findRecursiveFolderByValue.test.js
        - [x] findRecursiveFolderMetadatas.test.js
        - [x] formatBytes.test.js
        - [x] formatDuration.test.js
        - [x] getAllowedFiles.test.js
        - [x] getAPIInnerErrors.test.js
        - [x] getBreadcrumbDataCM.test.js
        - [x] getBreadcrumbDataML.test.js
        - [x] getFileExtension.test.js
        - [x] getFolderParents.test.js
        - [x] getFolderURL.test.js
        - [x] normalizeAPIError.test.js
        - [x] prefixFileUrlWithBackendUrl.test.js
        - [x] prefixPluginTranslations.test.js
        - [x] toSingularTypes.test.js
        - [x] typeFromMime.test.js
      - [x] appendSearchParamsToUrl.js
      - [x] containsAssetFilter.js
      - [x] createAssetUrl.js
      - [x] displayedFilters.js
      - [x] downloadFilte.js
      - [x] findRecursiveFolderByValue.js
      - [x] findRecursiveFolderMetadatas.js
      - [x] formatBytes.js
      - [x] formatDuration.js
      - [x] getAllowedFiles.js
      - [x] getAPIInnerErrors.js
      - [x] getBreadcrumbDataCM.js
      - [x] getBreadcrumbDataML.js
      - [x] getFileExtension.js
      - [x] getFolderParents.js
      - [x] getFolderURL.js
      - [x] getTrad.js
      - [x] index.js
      - [x] moveElement.js
      - [x] normalizeAPIError.js
      - [x] prefixFileUrlWithBackendUrl.js
      - [x] prefixPluginTranslations.js
      - [x] rawFileToAsset.js
      - [x] toSingularTypes.js
      - [x] typeFromMime.js
      - [x] urlsToAssets.js
      - [x] urlYupSchema.js
    - [ ] constants.js
    - [ ] index.js
    - [x] pluginId.js 
  - [x] tests  
    - [x] handlers.js
    - [x] setup.js
    - [x] utils.jsx 
- [ ] documentation 
  - [ ] content-api.json
- [x] .eslintignore
- [x] .eslintrc
- [x] .gitattributes
- [x] .gitignore
- [x] .npmignore
- [x] jest.config.front.js
- [x] jest.config.js
- [x] jsconfig.json
- [x] LICENSE
- [ ] package.json
- [ ] packup.config.ts
- [x] README.md
- [x] strapi-server.js 

Define the 
 - [x] Contracts

### CODE REVIEW

File: packages/core/upload/admin/src/components/MediaLibraryDialog/index.tsx [hunk 1]
```diff
@@ -12,9 +10,21 @@ const STEPS = {
   FolderCreate: 'FolderCreate',
 };
 
-export const MediaLibraryDialog = ({ onClose, onSelectAssets, allowedTypes }) => {
-  const [step, setStep] = useState(STEPS.AssetSelect);
-  const [folderId, setFolderId] = useState(null);
+import type { File } from '../../../../shared/contracts/files';
+
+interface MediaLibraryDialogProps {
+  allowedTypes?: string[];
```
Comment (remidej) [inline packages/core/upload/admin/src/components/MediaLibraryDialog/index.tsx:16]: you should inherit a more precise version of this type

File: packages/core/upload/admin/src/components/MediaLibraryInput/Carousel/CarouselAssets.tsx [hunk 2]
```diff
@@ -28,21 +55,21 @@ export const CarouselAssets = forwardRef(
...
-    const [isEditingAsset, setIsEditingAsset] = useState(false);
+    const [isEditingAsset, setIsEditingAsset] = React.useState(false);
 
     const currentAsset = assets[selectedAssetIndex];
 
     return (
       <>
         <CarouselInput
-          ref={forwardedRef}
+          ref={forwardedRef as React.LegacyRef<HTMLDivElement>}
```
Comment (remidej) [inline packages/core/upload/admin/src/components/MediaLibraryInput/Carousel/CarouselAssets.tsx:72]: why legacy?

File: packages/core/upload/admin/src/components/AssetDialog/BrowseStep/PaginationFooter/index.tsx
[the remark this answers is not in the document]
Comment (simotae14) [reply packages/core/upload/admin/src/components/AssetDialog/BrowseStep/PaginationFooter/index.tsx:1]: maybe you can add a ticket on our backlog as an enhancement, but I would stay focused on just the migration in this PR

File: packages/core/upload/admin/src/components/AssetDialog/BrowseStep/index.tsx [hunk 3]
```diff
@@ -100,8 +100,10 @@ export interface FileWithType extends File {
   type?: string;
 }
 
+type AllowedTypes = 'files' | 'images' | 'videos' | 'audios';
```
Comment (remidej) [inline packages/core/upload/admin/src/components/AssetDialog/BrowseStep/index.tsx:103]: You shouldn't have to define this type in multiple places (it's also in AssetCard), export/import it instead

File: packages/core/upload/admin/src/components/AssetDialog/index.tsx [hunk 4]
```diff
@@ -44,8 +44,10 @@ export interface FolderRow extends Folder {
   type?: string;
 }
 
+type AllowedTypes = 'files' | 'images' | 'videos' | 'audios';
```
Comment (remidej) [inline packages/core/upload/admin/src/components/AssetDialog/index.tsx:47]: same

File: packages/core/upload/admin/src/components/AssetGridList/index.tsx [hunk 5]
```diff
@@ -6,8 +6,10 @@ import { Draggable } from './Draggable';
 
 import type { File } from '../../../../shared/contracts/files';
 
+type AllowedTypes = 'files' | 'images' | 'videos' | 'audios';
```
Comment (remidej) [inline packages/core/upload/admin/src/components/AssetGridList/index.tsx:9]: same

File: packages/core/upload/admin/src/components/MediaLibraryDialog/index.tsx [hunk 6]
```diff
@@ -12,8 +12,10 @@ const STEPS = {
 
 import type { File } from '../../../../shared/contracts/files';
 
+type AllowedTypes = 'files' | 'images' | 'videos' | 'audios';
```
Comment (remidej) [inline packages/core/upload/admin/src/components/MediaLibraryDialog/index.tsx:15]: same

File: packages/core/upload/admin/src/components/MediaLibraryInput/index.tsx [hunk 7]
```diff
@@ -17,6 +17,8 @@ const STEPS = {
   FolderCreate: 'FolderCreate',
 };
 
+type AllowedTypes = 'files' | 'images' | 'videos' | 'audios';
```
Comment (remidej) [inline packages/core/upload/admin/src/components/MediaLibraryInput/index.tsx:20]: same

File: packages/core/upload/admin/src/components/TableList/index.tsx [hunk 8]
```diff
@@ -32,9 +32,11 @@ export interface FolderRow extends Folder {
   type?: string;
 }
 
+type AllowedTypes = 'files' | 'images' | 'videos' | 'audios';
```
Comment (remidej) [inline packages/core/upload/admin/src/components/TableList/index.tsx:35]: same

**Review by remidej — APPROVED** [review]
Tested 👍

**Review by Marc-Roig — APPROVED** [review]

**Comment by MarionLemaire** [comment]
Global test plan about ML so far : 

- [ ]  Add files
    - [ ]  Through Browse files
    - [ ]  Through Drag&Dop
    - [ ]  Check error messages (not a valid image, exceeding size)
- [ ]  Modify files
- [ ]  Delete files
- [ ]  Create folders
- [ ]  Move files from a folder to another one

**Comment by yanniskadiri** [comment]
Thanks for the feedback @MarionLemaire, I added these elements in the product backlog :)

**Pull request merged**